### PR TITLE
[Feat]: xlineutl snapshot status

### PR DESCRIPTION
**Pull request by themanforfree**
Please briefly answer these questions:
based on #615 
* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
support xlineutl snapshot status
* what changes does this pull request make?
support xlineutl snapshot status
* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
no

### CODE REVIEW

File: xlineutl/src/command/snapshot.rs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+async fn handle_status<P: AsRef<Path>>(snapshot_path: P) -> Result<()> {
+    let tempdir = tempdir()?;
+    let restore_rocks_engine = Engine::new(
+        EngineType::Rocks(tempdir.path().to_path_buf()),
+        &XLINE_TABLES,
+    )?;
+    restore_rocks_engine
+        .apply_snapshot_from_file(snapshot_path, &XLINE_TABLES)
+        .await?;
+    let mut status = Status::new();
```
Comment (Phoenix500526) [inline xlineutl/src/command/snapshot.rs:93]: You can directly use `default` here. There is no need to define a `new` method.

File: xline/src/storage/revision.rs [hunk 2]
```diff
@@ -46,12 +48,16 @@ impl Revision {
...
+    #[inline]
+    pub fn revision(&self) -> i64 {
         self.revision
     }
 
     /// Get sub revision
-    pub(crate) fn sub_revision(&self) -> i64 {
+    #[must_use]
+    #[inline]
+    pub fn sub_revision(&self) -> i64 {
```
Comment (Phoenix500526) [inline xline/src/storage/revision.rs:60]: Why remove `(crate)` of this `sub_revision` method?

**Review by Phoenix500526 — APPROVED** [review]
LGTM

File: crates/xlineutl/src/command/snapshot.rs [hunk 3]
```diff
@@ -39,3 +60,69 @@ async fn handle_restore<P: AsRef<Path>, D: Into<PathBuf>>(
...
+    for table in XLINE_TABLES {
+        hasher.write(table.as_bytes());
+        let kv_pairs = restore_rocks_engine.get_all(table)?;
+        let is_kv_table = table == KV_TABLE;
+        for (k, v) in kv_pairs {
+            hasher.write(k.as_slice());
+            hasher.write(v.as_slice());
+            if is_kv_table {
+                let rev = Revision::decode(k.as_slice());
+                status.revision = rev.revision();
```
Comment (bsbds) [inline crates/xlineutl/src/command/snapshot.rs:103]: This needs to do a max operation.

File: crates/xline/src/storage/revision.rs [hunk 4]
```diff
@@ -64,7 +68,9 @@ impl Revision {
     }
 
     /// Decode `Revision` from `&[u8]`
-    pub(crate) fn decode(mut buf: &[u8]) -> Self {
+    #[must_use]
+    #[inline]
+    pub fn decode(mut buf: &[u8]) -> Self {
         Self::new(buf.get_i64(), buf.get_i64())
     }
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply crates/xline/src/storage/revision.rs:75]: why? This decode won't panic or return Result

File: crates/xlineutl/src/command/snapshot.rs [hunk 5]
```diff
@@ -39,3 +60,69 @@ async fn handle_restore<P: AsRef<Path>, D: Into<PathBuf>>(
...
+
+/// handle restore snapshot to data dir
+#[inline]
+#[allow(clippy::arithmetic_side_effects)] // u64 is big enough
+async fn handle_status<P: AsRef<Path>>(snapshot_path: P) -> Result<()> {
+    let tempdir = tempdir()?;
+    let restore_rocks_engine = Engine::new(
+        EngineType::Rocks(tempdir.path().to_path_buf()),
+        &XLINE_TABLES,
+    )?;
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply crates/xlineutl/src/command/snapshot.rs:85]: memory engine's snapshot is memory-only implementation for testing, no need to recover from files

File: crates/xlineutl/src/command/snapshot.rs (hunk 3 above)
Comment (themanforfree) [reply crates/xlineutl/src/command/snapshot.rs:103]: Fixed

File: crates/xline/src/storage/revision.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (bsbds) [reply crates/xline/src/storage/revision.rs:75]: The `get_i64` could cause a panic

[the remark this answers is not in the document]
Comment (themanforfree) [reply crates/xline/src/storage/revision.rs:77]: Added

**Review by bsbds — APPROVED** [review]

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Pull request merged**